# Data cleaning 

This project focuses on cleaning the real-world audiobook dataset scraped from Audible. This messy dataset contains multiple inconsistencies such as wrong data types, redundant texts, mixed formats, and unstructured fields.

The objective of this project is to transform the dataset into a clean and structured format Python (pandas), making it suitable for further analysis and visualization.



## 1) Importing Libraries and Loading dataset

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv("datasets/audible_uncleaned.csv")
df.head()

,name,author,narrator,time,releasedate,language,stars,price
0,Geronimo Stilton #11 & #12,Writtenby:GeronimoStilton,Narratedby:BillLobely,2 hrs and 20 mins,04-08-08,English,5 out of 5 stars34 ratings,468.00
1,The Burning Maze,Writtenby:RickRiordan,Narratedby:RobbieDaymond,13 hrs and 8 mins,01-05-18,English,4.5 out of 5 stars41 ratings,820.00
2,The Deep End,Writtenby:JeffKinney,Narratedby:DanRussell,2 hrs and 3 mins,06-11-20,English,4.5 out of 5 stars38 ratings,410.00
3,Daughter of the Deep,Writtenby:RickRiordan,Narratedby:SoneelaNankani,11 hrs and 16 mins,05-10-21,English,4.5 out of 5 stars12 ratings,615.00
4,"The Lightning Thief: Percy Jackson, Book 1",Writtenby:RickRiordan,Narratedby:JesseBernstein,10 hrs,13-01-10,English,4.5 out of 5 stars181 ratings,820.00


## 2) Basic data exploration

In [3]:
df.sample(10)

,name,author,narrator,time,releasedate,language,stars,price
46973,Bulfinch's Mythology,Writtenby:ThomasBulfinch,Narratedby:FloGibson,13 hrs and 27 mins,23-01-09,English,Not rated yet,"1,218.00"
24117,Autobiografia di uno Yogi [Autobiography of a ...,Writtenby:ParamahansaYogananda,Narratedby:NormanMozzato,19 hrs and 44 mins,24-01-17,italian,Not rated yet,516.00
64282,Los 12 libros del Emperador Marco Aurelio [The...,Writtenby:EmperadorMarcoAurelioAntonio,Narratedby:SergioBustos,6 hrs and 30 mins,27-07-21,spanish,Not rated yet,234.00
73964,Shadows In The Wind,Writtenby:CarolynLampman,Narratedby:LaurieKlein,10 hrs and 57 mins,24-05-11,English,Not rated yet,750.00
47263,Free Billy,Writtenby:DonWinslow,Narratedby:EdHarris,1 hr and 6 mins,22-03-22,English,Not rated yet,501.00
26484,Charles Sobhraj,Writtenby:RaameshKoirala,Narratedby:UrvakshHoyvoy,6 hrs and 9 mins,08-02-19,English,5 out of 5 stars1 rating,428.00
7161,Meatballs Hairdos and Other New Year's Fun,Writtenby:LissaRovetch,Narratedby:HighlightsforChildren,5 mins,17-08-18,English,Not rated yet,46.00
10719,About Time,Writtenby:DavidRooney,Narratedby:DavidRooney,9 hrs and 58 mins,10-06-21,English,Not rated yet,888.00
44162,Hush,Writtenby:TalBauer,Narratedby:JoelLeslie,19 hrs and 23 mins,01-05-18,English,5 out of 5 stars1 rating,820.00
69706,Viaje a la Consciencia Eterna [Journey to Eter...,Writtenby:MarianoMenendez,Narratedby:MarianoMenendez,54 mins,01-02-22,spanish,Not rated yet,93.00


In [4]:
df.shape

(87489, 8)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87489 entries, 0 to 87488
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         87489 non-null  object
 1   author       87489 non-null  object
 2   narrator     87489 non-null  object
 3   time         87489 non-null  object
 4   releasedate  87489 non-null  object
 5   language     87489 non-null  object
 6   stars        87489 non-null  object
 7   price        87489 non-null  object
dtypes: object(8)
memory usage: 5.3+ MB


---
## 3) Data Summary and Key Issues in the data

- The dataset have 87489 rows and 8 columns.
- All columns of dataset are of object type.

#### Column Description
- `name` - audiobook title
- `author` - author name
- `narrator` - narrator name
- `time` - duration or length of the audiobook
- `releasedate` - release date of the audiobook
- `language` - language of the audiobook
- `stars` - rating of the audiobook and number of ratings
- `price` - price of the audiobook

### Key Issues in the Dataset


#### 1. Redundant Text in Columns
- The `author` and `narrator` columns contain unnecessary prefixes such as **Writtenby:** and **Narratedby:**.
- These prefixes do not add analytical value and need to be removed.

#### 2. Improper Name Formatting
- Author and narrator names are not properly formatted.There is not space between first and last name of author and narrator.
- This reduces readability and consistency.


#### 3. Incorrect Data Type (Time Column)
- The `time` column is stored as text (object type).
- It contains mixed formats like ""2 hrs and 20 mins""
- This is not suitable for any numerical analysis.


#### 4. Incorrect Date Format
- The `releasedate` column is stored as an object instead of datetime.

#### 5. Incorrect Data Type of price Column
- The `price` column is stored as text and contains commas.
- `price` columns should of float type.


#### 6. Multiple Values in stars column
- The `stars` column contain two pieces of information:
  - Rating (e.g., 4.5)
  - Number of ratings (e.g., 41 ratings)
- The column needs to be split into two sepertae columns.

#### 7. Columns are Stored as Object Type
- Every column in the dataset is of object type.
- We need to assign correct data type for these.

## 4) Data Cleaning

In [6]:
df.head()

,name,author,narrator,time,releasedate,language,stars,price
0,Geronimo Stilton #11 & #12,Writtenby:GeronimoStilton,Narratedby:BillLobely,2 hrs and 20 mins,04-08-08,English,5 out of 5 stars34 ratings,468.00
1,The Burning Maze,Writtenby:RickRiordan,Narratedby:RobbieDaymond,13 hrs and 8 mins,01-05-18,English,4.5 out of 5 stars41 ratings,820.00
2,The Deep End,Writtenby:JeffKinney,Narratedby:DanRussell,2 hrs and 3 mins,06-11-20,English,4.5 out of 5 stars38 ratings,410.00
3,Daughter of the Deep,Writtenby:RickRiordan,Narratedby:SoneelaNankani,11 hrs and 16 mins,05-10-21,English,4.5 out of 5 stars12 ratings,615.00
4,"The Lightning Thief: Percy Jackson, Book 1",Writtenby:RickRiordan,Narratedby:JesseBernstein,10 hrs,13-01-10,English,4.5 out of 5 stars181 ratings,820.00


#### 1) Cleaning 'author' column

In [7]:
# Removing 'Writtenby:' prefix from author column.
df["author"] = df["author"].str.replace("Writtenby:", "")

In [8]:
# Adding space between the first and last name
df["author"] = df["author"].str.replace(r"([a-z])([A-Z])", r"\1 \2", regex=True)

#### 2) Cleaning 'narrator' column

In [9]:
# Removing 'Narratedby:" prefix from narrator column.
df["narrator"] = df["narrator"].str.replace("Narratedby:", "")

In [10]:
# Adding space between narrator's first and last name
df["narrator"] = df["narrator"].str.replace(r"([a-z])([A-Z])", r"\1 \2", regex=True)

#### 3) Cleaning 'time' column

In [11]:
df["time"]

0         2 hrs and 20 mins
1         13 hrs and 8 mins
2          2 hrs and 3 mins
3        11 hrs and 16 mins
4                    10 hrs
                ...        
87484     7 hrs and 34 mins
87485     10 hrs and 7 mins
87486     19 hrs and 4 mins
87487      4 hrs and 8 mins
87488       6 hrs and 1 min
Name: time, Length: 87489, dtype: object

In [12]:
# Extracting hours from time column
df["hours"] = df["time"].str.extract(r"(\d+)\s*hr[s]?").fillna(0).astype(int)

In [13]:
# Extacting minutes from time column
df["minutes"] = df["time"].str.extract(r"(\d+)\s*min[s]?").fillna(0).astype(int)

In [14]:
# Creating new column 'duration(mins)' 
df["duration(min)"] = df["hours"] * 60 + df["minutes"]

#### 4) Converting 'date' column to datetime type 

In [15]:
df["releasedate"] = pd.to_datetime(df["releasedate"], format="%d-%m-%y")

#### 5) Cleaning 'stars' column

In [16]:
# Creating new column 'ratings' from stars
df["ratings"] = df["stars"].str.extract(r"(\d(?:\.\d)?) out").astype(float)

In [17]:
# Creating new columnn 'numberofratings" from stras
df["numberofratings"] = df["stars"].str.extract(r"(\d+) rating").fillna(0).astype("int64")

#### 6) Cleaning 'price' column

In [22]:
# 338 records in the price column are labeled as 'Free', we can replace this with 0
df[df["price"] == "Free"].shape

(338, 13)

In [29]:
df["price"] = df["price"].str.replace("Free", "0").str.replace(",", "").astype(float)

In [30]:
df.head()

,name,author,narrator,time,releasedate,language,stars,price,hours,minutes,duration(min),ratings,numberofratings
0,Geronimo Stilton #11 & #12,Geronimo Stilton,Bill Lobely,2 hrs and 20 mins,2008-08-04,English,5 out of 5 stars34 ratings,468.0,2,20,140,5.0,34
1,The Burning Maze,Rick Riordan,Robbie Daymond,13 hrs and 8 mins,2018-05-01,English,4.5 out of 5 stars41 ratings,820.0,13,8,788,4.5,41
2,The Deep End,Jeff Kinney,Dan Russell,2 hrs and 3 mins,2020-11-06,English,4.5 out of 5 stars38 ratings,410.0,2,3,123,4.5,38
3,Daughter of the Deep,Rick Riordan,Soneela Nankani,11 hrs and 16 mins,2021-10-05,English,4.5 out of 5 stars12 ratings,615.0,11,16,676,4.5,12
4,"The Lightning Thief: Percy Jackson, Book 1",Rick Riordan,Jesse Bernstein,10 hrs,2010-01-13,English,4.5 out of 5 stars181 ratings,820.0,10,0,600,4.5,181


#### 7) Droping unwanted columns and rearranging columns

In [35]:
df = df[["name", "author", "releasedate", "time", "duration(min)",
         "language", "ratings", "numberofratings", "price"]]

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87489 entries, 0 to 87488
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   name             87489 non-null  object        
 1   author           87489 non-null  object        
 2   releasedate      87489 non-null  datetime64[ns]
 3   time             87489 non-null  object        
 4   duration(min)    87489 non-null  int64         
 5   language         87489 non-null  object        
 6   ratings          15072 non-null  float64       
 7   numberofratings  87489 non-null  int64         
 8   price            87489 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(2), object(4)
memory usage: 6.0+ MB


#### 8) Exporting cleaned dataframe as csv

In [38]:
df.to_csv("datasets/audible_cleaned.csv", index=False)